### PR TITLE
feat: migrations can be provided as a map of loaded modules

### PR DESCRIPTION
**Pull request by polad**
Currently it's not possible to use `node-migrate` with packagers relying on static analysis (e.g. webpack) because they need to know all module and file dependencies during bundling. Unfortunately the `node-migrate` package is dynamically loading migration files in `lib/load-migrations.js` using `require(filepath)` where `filepath` is a variable hence webpack cannot infer during compile time what folders and files to include for bundling in the final output.

This small PR solves the problem by introducing a new `migrationsMap` property in the options that are provided to the `migrate.load(opts, cb)` function like so:
```javascript
const migrationsMap = {
  "1673460432335-create-users-table.js": require("../migrations/1673460432335-create-users-table"),
  "1675438233157-create-sales-table.js": require("../migrations/1675438233157-create-sales-table"),
}

migrate.load({ migrationsMap }, (err, set) => {
  // you callback code here
})
```
This way there is no need to scan the migrations directory to include the migration files dynamically since they are statically included in the migrationsMap so they will be included by the bundlers like webpack.

The `migrationsDirectory` is ignored for loading files when the `migrationsMap` option is provded.

### CODE REVIEW

**Comment by wesleytodd** [comment]
Ah, interesting idea. So I want to say first, while I have in the past used a bundler on server code I don't think it is a good practice. So if we add this I don't think we should do so *just* because of the bundler use case. That said, I think a strategy to define migrations without dynamic filesystem require/imports is good. Technically you can do it already by using the library exports and not the cli, but I don't think that is very ergonomic.

I think the main thing that is missing here now is sorting. Because map keys are not sorted, to get the "same behavior" for ordering you need to sort the keys before adding them. And two nit picks:

1. Now that we would have two ways to load the migrations I would like to see them broken out into separate functions.
2. Can we just name the option `migrations`?

**Comment by wesleytodd** [comment]
> If you're asking when we run migrate lib

Yeah that is what I was asking. Because bundling this code only makes sense if you need to run it in the lambda. Anyway, not a big deal and I agree this is a good feature.

> Objects are Maps

Yes, but now that there is an actual implementation called map, using that as a suffix typically implies it should be a `Map` instance. Since this is not that I don't see a reason to include the suffix.

> Can we merge the PR?

Taking a look, but yeah I think it should be good.

File: lib/load-migrations.js [hunk 1]
```diff
@@ -16,8 +53,13 @@ function loadMigrationsIntoSet (options, fn) {
   const set = opts.set
   const store = opts.store
   const ignoreMissing = !!opts.ignoreMissing
+  const migrations = opts.migrations
   const migrationsDirectory = path.resolve(opts.migrationsDirectory || 'migrations')
+
+  // filterFn :: String -> boolean
   const filterFn = opts.filterFunction || (() => true)
```
Comment (wesleytodd) [inline lib/load-migrations.js:60]: Can we remove these unrelated changes?

File: Readme.md [hunk 2]
```diff
@@ -230,6 +230,7 @@ Calls the callback with a `Set` based on the options passed.  Options:
 
 - `set`: A set instance if you created your own
 - `stateStore`: A store instance to load and store migration state, or a string which is a path to the migration state file
+- `migrations`: An object where keys are migration file names and values are migration modules loaded with `require()`
```
Comment (wesleytodd) [inline Readme.md:233]:
```suggestion
- `migrations`: An object where keys are migration names and values are migration objects
```

I think they can be loaded however you want, you can define them inline if you want. I don't think we need to specify "with require"

File: lib/load-migrations.js [hunk 3]
```diff
@@ -7,6 +7,43 @@ const Migration = require('./migration')
 
 module.exports = loadMigrationsIntoSet
 
+async function loadFromMigrationsDirectory (migrationsDirectory, filterFn) {
```
Comment (wesleytodd) [inline lib/load-migrations.js:10]: I notice a few unrelated changes made in here. Was there a reason to make changes and not just copy them out? Ideally we would just copy them out as is. I would not block on this, but for future reference, keeping PRs to the minimal related changes is best so we don't have to spend time digging around for what changed.

**Review by wesleytodd — CHANGES_REQUESTED** [review]

**Comment by wesleytodd** [comment]
Made a couple of final comments. But yeah, other than those I can merge and release this.

File: lib/load-migrations.js (hunk 3 above)
Comment (polad) [reply lib/load-migrations.js:10]: Previously this part would modify the `migMap` object/map ;) as well as maintain an array of `Migration` objects. So I had to refactor this part since it's extracted into a function now and should have one return output, which is `migMap` object.

File: Readme.md (hunk 2 above)
Comment (polad) [reply Readme.md:233]: Fair point.

File: lib/load-migrations.js (hunk 1 above)
Comment (polad) [reply lib/load-migrations.js:60]: I added these comments because it's hard to tell that the signatures of the `sortFunction` and `filterFunction` are different. `sortFunction` accepts a `Migration` as an argument but the `filterFunction` takes a `String` title of the migration. Perhaps they are better off in the `Readme` file. I will remove these...

File: lib/load-migrations.js (hunk 3 above)
Comment (wesleytodd) [reply lib/load-migrations.js:10]: oh hm. I was talking about the inline'ing of line 30. Maybe it is good, just meant I had to read each line in this method instead of just skimming it as "copied". Anyway, will mark this as resolved.

File: lib/load-migrations.js (hunk 1 above)
Comment (wesleytodd) [reply lib/load-migrations.js:62]: I would love it if sonmeone wanted to add some JSDoc to these. Would be nice to generate some types instead of having to maintain those manually.

**Comment by polad** [comment]
@wesleytodd I applied you suggestions in my latest commits. Let me know if there is anything else.

Thanks again

**Comment by wesleytodd** [comment]
Hey, sorry busy day. I probably wont be able to get to this tonight but will try and look in the morning.

**Comment by polad** [comment]
Hey @wesleytodd :) whenever you get a chance to take another look. Thanks 👍

**Comment by polad** [comment]
hey @wesleytodd just bumping this PR as a friendly reminder :) there are few minor changes really. Thanks

**Comment by wesleytodd** [comment]
Sorry looks like I dropped the ball on this. Taking a look now.

**Review by wesleytodd — APPROVED** [review]

**Comment by wesleytodd** [comment]
Looks like I still can't change the settings to allow the actions to run on this PR. @tj it would be great if you could make me an administrator on this repo. I almost moved it into my account once before because of this, maybe I should just do that.  I will pull it down and manually run the tests and stuff I guess.

**Comment by wesleytodd** [comment]
Released as 2.1.0
